pallets/loans/src/types/cashflow.rs

```diff
@@ -0,0 +1,451 @@
+pub enum PayDownSchedule {
+	/// No restrictions on how the paydown should be done.
+	None,
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub struct CashflowPayment<Balance> {
+	pub when: Seconds,
+	pub principal: Balance,
+	pub interest: Balance,
+}
+
+/// Specify the repayment schedule of the loan
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub struct RepaymentSchedule {
+	/// Expected repayments date for remaining debt
+	pub maturity: Maturity,
+
+	/// Period at which interest is paid
+	pub interest_payments: InterestPayments,
+
+	/// How much of the initially borrowed amount is paid back during interest
+	/// payments
+	pub pay_down_schedule: PayDownSchedule,
+}
+
+impl RepaymentSchedule {
+	pub fn is_valid(&self, now: Seconds) -> Result<bool, DispatchError> {
+		let valid = match self.interest_payments {
+			InterestPayments::None => true,
+			InterestPayments::Monthly(_) => {
+				match self.maturity.date() {
+					Some(maturity) => {
+						let start = date::from_seconds(now)?;
+						let end = date::from_seconds(maturity)?;
+
+						// We want to avoid creating a loan with a cashflow consuming a lot of
+						// computing time Maximum 40 years, which means a cashflow list of 40 * 12
+						// elements
+						end.year() - start.year() <= 40
+					}
+					None => false,
+				}
+			}
+		};
+
+		Ok(valid && self.maturity.is_valid(now))
+	}
+
+	pub fn generate_cashflows<Balance, Rate>(
+		&self,
+		origination_date: Seconds,
+		principal: Balance,
+		interest_rate: &InterestRate<Rate>,
+	) -> Result<Vec<CashflowPayment<Balance>>, DispatchError>
+	where
+		Balance: FixedPointOperand + EnsureAdd + EnsureSub + std::fmt::Debug,
+		Rate: FixedPointNumber,
+	{
+		let Some(maturity) = self.maturity.date() else {
+			return Ok(Vec::new());
+		};
+
+		let start_date = date::from_seconds(origination_date)?;
+		let end_date = date::from_seconds(maturity)?;
+
+		let timeflow = match &self.interest_payments {
+			InterestPayments::None => vec![],
+			InterestPayments::Monthly(reference_day) => {
+				date::monthly_intervals(start_date, end_date, (*reference_day).into())?
+			}
+		};
+
+		let total_weight = timeflow
+			.iter()
+			.map(|(_, weight)| weight)
+			.try_fold(0, |a, b| a.ensure_add(*b))?;
+
+		let lifetime = maturity.ensure_sub(origination_date)?.ensure_into()?;
+		let interest_rate_per_lifetime = ensure_pow(interest_rate.per_sec()?, lifetime)?;
+		let interest_at_maturity = interest_rate_per_lifetime
+			.ensure_mul_int(principal)?
+			.ensure_sub(principal)?;
+
+		timeflow
+			.into_iter()
+			.map(|(date, weight)| {
+				let proportion = FixedU128::ensure_from_rational(weight, total_weight)?;
+				let principal = proportion.ensure_mul_int(principal)?;
+				let interest = proportion.ensure_mul_int(interest_at_maturity)?;
+
+				Ok(CashflowPayment {
+					when: date::into_seconds(date)?,
+					principal,
+					interest,
+				})
+			})
+			.collect()
```

Comment:
If `InterestPayments::None` is choosen, there will be no `CashflowPayment`. 

But rather we want the `InterestPayments::None` to have a single payment at the mutrity date. Also, we could think about renaming `None` to `AtMaturity`.

Reply:
Mmmm... Makes sense. Doing it 👍🏻

Reply:
WDYT `OnceAtMaturity` to express that you only have one entry in that case?

Reply:
Done here: [f9db4fc](https://github.com/centrifuge/centrifuge-chain/pull/1797/commits/f9db4fcfe2211c47133005b0640df5437000235c)

Reply:
BTW, the expected interest to pay in the cashflow will be different from the expected interest to pay with external loans. Are we ok with that?

Reply:
Why?

Reply:
Because the interest in external pricing loan is affected by the notional price.

Reply:
https://github.com/centrifuge/centrifuge-chain/blob/main/pallets/loans/src/entities/pricing/external.rs#L220-L226

Line `225` is equal to how we compute the interest in the `cashflows`. That contains the principal + interest. 

Nevertheless, in the cashflows & internal pricing, we subtract that by the outstanding principal, and for external pricing, the outstanding principal is based on the notional

(BTW, we should unify these 3 points of computing the interest)

Reply:
I think I have a fast fix to unify them. Working on this

Reply:
https://github.com/centrifuge/centrifuge-chain/pull/1864

Reply:
> If InterestPayments::None is choosen, there will be no CashflowPayment.

I still think that this is not working as expected. Would like to see a test where the actual `Payment` is shown made for externally priced assets.

Reply:
> If InterestPayments::None is choosen, there will be no CashflowPayment.

This is working as expected because now, for InterestPayments, we're returning one element in the cashflow. I added a test.

What is not working as expected is the external price. But with the fix above it should. Still pending the test. Will do later today or early tomorrow